### PR TITLE
Restricting coarse element ID assignment to fine mesh subdomains

### LINKED ISSUE #23708: Subdomain restriction for CoarseMeshExtraElementIDGenerator
## Reason

It would be useful to assign extra element IDs based on two different coarse meshes. For instance, creating a coarse mesh for CMFD where pin resolved regions are smoothed out, but control rod/drum regions are retain the fine mesh structure. Piecing these regions together in one mesh would be difficult due to inevitable hanging nodes. However, CMFD doesn't care about nodes at all and having a fully conforming coarse mesh is unnecessary. Assigning the coarse IDs by overlaying two different coarse meshes would resolve this difficulty.

## Design

Easiest way I can think of is restricting the element ID assignment on subdomains with a new parameter `subdomains`. When empty, the entire fine mesh is assigned IDs. When it is specified, only subdomains requested are assigned IDs.

## Impact

Even MORE flexible coarse mesh definition.

### PR DESCRIPTION
**Pull request by zachmprince**
Closes #23708

### CODE REVIEW

File: modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C [hunk 1]
```diff
@@ -53,11 +57,32 @@ CoarseMeshExtraElementIDGenerator::generate()
...
+  // If we are not going through the entire mesh and the ID already exists, we need to offset the
+  // assigned ID so that we don't copy ones that are already there
+  dof_id_type id_offset = 0;
+  if (already_has_id && isParamValid("subdomains"))
+  {
+    for (auto & elem : mesh->active_element_ptr_range())
+    {
+      dof_id_type elem_id = elem->get_extra_integer(coarse_id);
+      if (elem_id != DofObject::invalid_id && elem_id > id_offset)
+        id_offset = elem_id;
```
Comment (YaqiWang) [inline modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C:75]: Better to add a note in the markdown file on this.

File: modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C [hunk 2]
```diff
@@ -53,11 +57,32 @@ CoarseMeshExtraElementIDGenerator::generate()
...
+  const bool already_has_id = mesh->has_elem_integer(_coarse_id_name);
+  if (!already_has_id)
     coarse_id = mesh->add_elem_integer(_coarse_id_name);
   else
     coarse_id = mesh->get_elem_integer_index(_coarse_id_name);
 
+  // If we are not going through the entire mesh and the ID already exists, we need to offset the
+  // assigned ID so that we don't copy ones that are already there
+  dof_id_type id_offset = 0;
+  if (already_has_id && isParamValid("subdomains"))
```
Comment (YaqiWang) [inline modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C:69]: `isParamValid("subdomains")` here will always return true, unless remove its default value in the validParams. In that case, Line 83 will need to be protected.

File: modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C [hunk 3]
```diff
@@ -53,11 +57,32 @@ CoarseMeshExtraElementIDGenerator::generate()
...
+    {
+      dof_id_type elem_id = elem->get_extra_integer(coarse_id);
+      if (elem_id != DofObject::invalid_id && elem_id > id_offset)
+        id_offset = elem_id;
+    }
+    id_offset++;
+  }
+
+  // Get the requested subdomain IDs
+  std::set<SubdomainID> include_subdomains;
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C:81]:
```suggestion
  std::set<SubdomainID> included_subdomains;
```

File: modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C [hunk 4]
```diff
@@ -53,11 +57,32 @@ CoarseMeshExtraElementIDGenerator::generate()
...
+  for (const auto & sid :
+       MooseMeshUtils::getSubdomainIDs(*mesh, getParam<std::vector<SubdomainName>>("subdomains")))
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C:83]: does this fail when the subdomain is not in the mesh ? (due to a typo for example?)

File: modules/reactor/test/tests/meshgenerators/coarse_mesh_extra_element_id_generator/tests [hunk 5]
```diff
@@ -26,6 +26,15 @@
     issues = '#19297'
     recover = false
   []
+  [coarse_element_restrict_subdomain]
+    type = Exodiff
+    input = coarse_elem_restrict_subdomain.i
+    exodiff = coarse_elem_restrict_subdomain_out.e
+    requirement = 'The system shall include the ability to assign extra element IDs based on a coarse mesh for subdomain-restricted elements of a fine mesh.'
```
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/coarse_mesh_extra_element_id_generator/tests:33]:
```suggestion
    requirement = 'The system shall be able to assign extra element IDs based on a coarse mesh for subdomain-restricted elements of a fine mesh.'
```

File: modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C (hunk 4 above)
Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/CoarseMeshExtraElementIDGenerator.C:83]: ok it doesnt it just gives invalid_subdomain id

we need a check that are subdomain names passed are in the mesh

**Review by GiudGiud — CHANGES_REQUESTED** [review]

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
@zachmprince this broke next btw

https://civet.inl.gov/event/117135/

**Comment by lindsayad** [comment]
it's almost more red than green 😆

**Pull request merged**